add(application): migration from ghcr to gar (#564)
nais/doc: docs/nais-application/migration.md

> # Migrate from GHCR to GAR

Reviewer: ```suggestion
# Migrate from GitHub Container Registry (GHCR) to Google Artifact Registry (GAR)
```

> 
> ## Introduction
> 
> This document describes how to migrate from GHCR (GitHub Container Registry) to GAR (Google Artifact Registry).
> 

Reviewer: ```suggestion
```

## Why migrate?

> GHCR serves as a container registry hosted by GitHub, but it has certain limitations, such as imposing strict rate

Reviewer: container registry or "OCI registry", Maybe it's long term beneficial to point out that you can put any oci blobs in there.

limits on the number of requests. We've encountered challenges with these rate limits, particularly when upgrading a
cluster or restoring from a backup. In such instances, we've had to patiently wait for the rate limits to reset before
deploying / restoring our applications.

On the other hand, GAR is a container registry hosted by Google, and we find it to be a more favorable solution for our
needs. Additionally, as we transition to the cloud, the seamless integration with the Google Cloud Platform further
enhances its appeal as a beneficial option for our operations.

### Migrate

> If you currently have a workflow utilizing GHCR, transitioning to GAR can be achieved by updating the workflow with the
> following steps, to illustrate the migration process, let's consider an example of a workflow that currently relies on
> GHCR. We will use the example as a case study to demonstrate how to migrate to GAR.

Reviewer: ```suggestion
You can switch from GHCR to GAR by following these steps
to update your current workflow. Let's simplify this by taking an
example of a workflow using GHCR and show you exactly how to change it
to GAR.
```

or something like that, This section has pretty fancy words and a passive voice, idk how we tone-of-voice that.

The [docker-build-action](https://github.com/nais/docker-build-push) is employed for building and publishing the

> Docker image to GHCR.

Reviewer 2: I think you mean GAR?

Author: yepp :)

      - uses: nais/deploy/actions/deploy@v1
        env:
          APIKEY: ${{ secrets.NAIS_DEPLOY_APIKEY }}
          CLUSTER: target-cluster
          RESOURCE: my-resource/.nais/nais-cluster.yaml
          VAR: image=${{ env.IMAGE }}
```

Initially, eliminate the environment variable: `IMAGE: ghcr.io/navikt/my-team/my-app:${{ github.sha }}`, as
the `nais/docker-build-action`, automatically generates this value and outputs it as the `image` variable.

```yaml
europe-north1-docker.pkg.dev/nais-management-233d/my-team/my-app
```

The resulting image reference is composed of the following:

* Registry: `europe-north1-docker.pkg.dev`
* Google Cloud Project: `nais-management-233d`
* Team: `my-team`
* Application: `my-app`

The tag is automatically generated by the `nais/docker-build-action` or can be specified as input to
the [action](https://github.com/nais/docker-build-push).

If you wish to refer to the image later, please consult the guide
on [Image registry](https://doc.nais.io/guides/application/#step-6-push-your-image-to-our-image-registry).

* Substitute the preceding code block and replace it with the following:

```yaml
  - name: Build and publish Docker image

>     if: ithub.ref_name == 'main'

Reviewer: ```suggestion
    if: github.ref_name == 'main'
```

    run: |
      echo ${GITHUB_TOKEN} | docker login docker.pkg.github.com -u ${GITHUB_REPOSITORY} --password-stdin
      echo ${GITHUB_TOKEN} | docker login ghcr.io -u ${{github.actor}} --password-stdin
      docker build -t ${IMAGE} .
      docker push ${IMAGE}
```

```yaml
  - name: Push docker image to GAR and sign image
    uses: nais/docker-build-push@v0
    id: docker-build-push
    with:
      team: my-team
      identity_provider: ${{ secrets.NAIS_WORKLOAD_IDENTITY_PROVIDER }} # Provided as Organization Secret
      project_id: ${{ vars.NAIS_MANAGEMENT_PROJECT_ID }} # Provided as Organization Variable
```

* In the "deploy" step, substitute the given image reference, with the output of the previous step. The output of the
  previous step is the `image` variable, given the example over it can be referenced as
  follows; `${{ steps.docker-build-push.outputs.image }}`. The id `docker-build-push` is the id of the previous step,
  there is where your new image will be outputted. In this example we devided our workflow into two jobs, `build`
  and `deploy`. The `deploy` job depends on the `build` job, and will not run unless the `build` job is successful.

>   The `build` job outputs the image reference as `image`, and the `deploy` job can reference this output

Reviewer 2: This part hasn't been shown yet, it is only shown in the final workflow at the bottom. That can be confusing.

  as `needs.build.outputs.image`.

```yaml
    env:
      APIKEY: ${{ secrets.NAIS_DEPLOY_APIKEY }}
      CLUSTER: dev-gcp
      RESOURCE: my-resource/.nais/nais-prod.yaml
      VAR: image=${{ env.IMAGE }}
```

```yaml
    env:
      APIKEY: ${{ secrets.NAIS_DEPLOY_APIKEY }}
      CLUSTER: dev-gcp
      RESOURCE: my-resource/.nais/nais-prod.yaml
      VAR: image=${{ needs.build.outputs.image }}
```

> * If you haven't already, add scoped permissions to the workflow. Afterward, modify the following:

Reviewer: Add a section on why this is needed?

Author: agree
